app/models/task.rb

```diff
@@ -140,15 +140,26 @@ def initialize_derivate(user = nil)
   end
 
   def average_rating
-    if ratings.empty?
-      0
-    else
-      ratings.sum(&:rating).to_f / ratings.size
-    end
+    return @average_rating if @average_rating
+
+    category_averages = Rating::CATEGORIES.map {|category| "AVG (#{category}) AS #{category}" }.join(',')
+
+    @average_rating = ActiveRecord::Base.connection.exec_query(
+      "SELECT #{category_averages} FROM ratings WHERE ratings.task_id = $1",
+      'Averaging ratings', [id]
+    ).first.symbolize_keys.transform_values {|rating| (rating || 0).round(1) }
+  end
+
+  def average_rating_stars
+    @average_rating_stars ||= average_rating.transform_values {|rating| (rating * 2).round / 2.0 }
+  end
```

Comment:
Manually concatenating an SQL string is working here and, most likely, secure. Still, I am usually not a big fan of that, since one must be very careful when changing the query later on. One potential alternative that is definitely secure (but unfortunately longer):

```rb
def average_rating # rubocop:disable Metrics/AbcSize
  return @average_rating if @average_rating

  ratings_arel = Rating.arel_table

  category_averages = Rating::CATEGORIES.map do |category|
    average_rating = ratings_arel[category].average
    coalesced_rating = Arel::Nodes::NamedFunction.new('COALESCE', [average_rating, 0])
    rounded_rating = Arel::Nodes::NamedFunction.new('ROUND', [coalesced_rating, 1])
    rounded_rating.as(category.to_s)
  end

  condition = ratings_arel[:task_id].eq(id)
  query = ratings_arel.project(category_averages).where(condition)
  result = ActiveRecord::Base.connection.exec_query(query.to_sql, 'Averaging ratings').first

  @average_rating = result.symbolize_keys
end
```